### PR TITLE
[xray-exporter] Add ability to export cucumber test cases

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/converter/CucumberScenarioConverter.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        return scenario.collectSteps()
+                       .stream()
+                       .map(Step::getValue)
+                       .map(processStep())
+                       .collect(Collectors.collectingAndThen(Collectors.toList(), beforeJoining))
+                       .stream()
+                       .collect(Collectors.joining(lineSeparator()));
+    }
+
+    private static UnaryOperator<String> processStep()
```
Comment (ikalinin1) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/converter/CucumberScenarioConverter.java:73]: private static String processStep(String step) and CucumberScenarioConverter::processStep 
?

Comment (ikalinin1) [reply vividus-to-xray-exporter/src/main/java/org/vividus/xray/converter/CucumberScenarioConverter.java:73]: and it seems like replaceCommentSign

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import org.vividus.bdd.model.jbehave.Scenario;
+import org.vividus.bdd.model.jbehave.Step;
+
+public enum TestCaseType
+{
+    MANUAL, CUCUMBER;
+
+    public static TestCaseType getTestCaseType(Scenario scenario)
+    {
+        return !scenario.collectSteps().stream().map(Step::getOutcome).allMatch("comment"::equals)
```
Comment (ikalinin1) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java:29]: noneMatch

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/util/StoryUtils.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            return scenarios;
+        }
+
+        Parameters lifecycleParameters = lifecycle.getParameters();
+        int partition = scenarios.size() <= 1 ? 1 : scenarios.size() / lifecycleParameters.getValues().size();
+        List<List<Scenario>> scenariosPerExample = ListUtils.partition(scenarios, partition);
+
+        return IntStream.range(0, partition)
+                .mapToObj(index -> scenariosPerExample.stream()
```
Comment (ikalinin1) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/util/StoryUtils.java:54]: don't you mind to align stream? :)

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/databind/AbstractTestCaseSerializer.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        generator.writeEndObject();
+    }
+
+    protected void writeObjectWithValueField(JsonGenerator generator, String objectKey, String fieldValue)
+            throws IOException
+    {
+        writeObjectWithField(generator, objectKey, "value", fieldValue);
+    }
+
+    protected abstract void customSerialize(T testCase, JsonGenerator generator) throws IOException;
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/databind/AbstractTestCaseSerializer.java:72]: `customSerialize` -> `serializeCustomFields`

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java [hunk 5]
```diff
@@ -115,17 +142,36 @@ public void exportResults() throws IOException
...
+        parameters.setScenarioType(cucumberScenario.getType());
+        parameters.setScenario(cucumberScenario.getScenario());
+        return parameters;
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T extends AbstractTestCaseParameters> T createTestCaseParameters(TestCaseType type, Scenario scenario,
+            Supplier<T> parametersFactory)
+    {
+        AbstractTestCaseParameters parameters = parametersFactory.get();
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java:168]: `T parameters = parametersFactory.get();`

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java [hunk 6]
```diff
@@ -93,16 +116,20 @@ public void exportResults() throws IOException
...
         try
         {
-            TestCaseParameters parameters = createTestCaseParameters(storyTitle, scenario);
+            TestCaseType testCaseType = TestCaseType.getTestCaseType(scenario);
 
             String testCaseId = ensureOneValueOrNull(scenarioMeta, "testCaseId");
+
+            CreateParametersFunction parameters = parameterFactories.get(testCaseType);
+            AbstractTestCase testCase = testCaseFactories.get(testCaseType)
+                    .apply(parameters.apply(scenarioTitle, scenario));
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java:125]:
```java
AbstractTestCaseParameters parameters = parameterFactories.get(testCaseType).apply(scenarioTitle, scenario);
AbstractTestCase testCase = testCaseFactories.get(testCaseType).apply(parameters);
```

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/factory/TestCaseFactory.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        AbstractTestCase testCase = testCaseFactory.get();
+        testCase.setType(parameters.getType().getValue());
+        testCase.setProjectKey(projectKey);
+        testCase.setAssignee(assignee);
+        testCase.setSummary(parameters.getSummary());
+        testCase.setLabels(parameters.getLabels());
+        testCase.setComponents(parameters.getComponents());
+        return (T) testCase;
+    }
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/factory/TestCaseFactory.java:66]:
```java
    public ManualTestCase createManualTestCase(ManualTestCaseParameters parameters)
    {
        ManualTestCase testCase = new ManualTestCase();
        fillTestCase(parameters, testCase);
        testCase.setManualTestSteps(parameters.getSteps());
        return testCase;
    }

    public CucumberTestCase createCucumberTestCase(CucumberTestCaseParameters parameters)
    {
        CucumberTestCase testCase = new CucumberTestCase();
        fillTestCase(parameters, cucumberTestCase);
        testCase.setScenarioType(parameters.getScenarioType());
        testCase.setScenario(parameters.getScenario());
        return testCase;
    }

    private void fillTestCase(AbstractTestCaseParameters parameters, AbstractTestCase testCase)
    {
        testCase.setType(parameters.getType().getValue());
        testCase.setProjectKey(projectKey);
        testCase.setAssignee(assignee);
        testCase.setSummary(parameters.getSummary());
        testCase.setLabels(parameters.getLabels());
        testCase.setComponents(parameters.getComponents());
    }
```

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import org.vividus.bdd.model.jbehave.Scenario;
+import org.vividus.bdd.model.jbehave.Step;
+
+public enum TestCaseType
+{
+    MANUAL, CUCUMBER;
+
+    public static TestCaseType getTestCaseType(Scenario scenario)
+    {
+        return scenario.collectSteps().stream().map(Step::getOutcome).noneMatch("comment"::equals)
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java:29]: could you please explain the logic?

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/util/StoryUtils.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.vividus.bdd.model.jbehave.Scenario;
+import org.vividus.bdd.model.jbehave.Story;
+
+public final class StoryUtils
+{
+    private StoryUtils()
+    {
+    }
+
+    public static List<Scenario> getFoldScenarios(Story story)
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/util/StoryUtils.java:39]: `getFoldedScenarios`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java (hunk 8 above)
Comment (uarlouski) [reply vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java]: That was a bit incorrect, now fixed and tests were added

File: docs/modules/integrations/pages/xray-exporter.adoc [hunk 10]
```diff
@@ -60,6 +61,12 @@ The Xray is a Jira plugin that uses custom Jira fields for it's data, one of the
 |`jira.fields-mapping.manual-steps`
 |Key of a field containing collection of manual steps
 
+|`jira.fields-mapping.cucumber-scenario-type`
+|Key of a field containing cucumer scenario type either `Scenario` or `Scenario Outline`
```
[the remark this answers is not in the document]
Comment (uarlouski) [reply docs/modules/integrations/pages/xray-exporter.adoc:64]: in progress

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java (hunk 2 above)
Comment (uarlouski) [reply vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java:29]: cucumber scenario is allowed to have comments

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/converter/CucumberScenarioConverter.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    {
+    }
+
+    public static CucumberScenario convert(Scenario scenario)
+    {
+        if (scenario.getExamples() != null)
+        {
+            return new CucumberScenario("Scenario Outline", buildExampleScenario(scenario));
+        }
+        return new CucumberScenario("Scenario", getSteps(scenario));
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/converter/CucumberScenarioConverter.java:43]: `getSteps(scenario)` can be moved to a variable, placed before `if` and then re-used twice

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+public enum TestCaseType
+{
+    MANUAL, CUCUMBER;
+
+    public static TestCaseType getTestCaseType(Scenario scenario)
+    {
+        return !scenario.collectSteps().stream().map(Step::getOutcome).allMatch("comment"::equals)
+                   ? CUCUMBER
+                   : MANUAL;
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/model/TestCaseType.java:31]:
```java
return scenario.collectSteps().stream().map(Step::getOutcome).allMatch("comment"::equals)
                   ? MANUAL
                   : CUCUMBER;
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/integrations/pages/xray-exporter.adoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (uarlouski) [reply docs/modules/integrations/pages/xray-exporter.adoc]: added additional pictures of manual and cucumber cases to make mappings visually clear to end users

**Review by ikalinin1 — APPROVED** [review]

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java [hunk 13]
```diff
@@ -115,16 +141,34 @@ public void exportResults() throws IOException
...
     }
 
-    private TestCaseParameters createTestCaseParameters(String storyTitle, Scenario scenario) throws SyntaxException
+    private ManualTestCaseParameters createManualTestCaseParameters(String storyTitle, Scenario scenario)
+            throws SyntaxException
     {
-        String scenarioTitle = scenario.getTitle();
-        List<Meta> scenarioMeta = scenario.getMeta();
+        ManualTestCaseParameters parameters = createTestCaseParameters(TestCaseType.MANUAL, scenario,
+                ManualTestCaseParameters::new);
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java:148]:
```java 
ManualTestCaseParameters parameters = new ManualTestCaseParameters();
fillTestCaseParameters(parameters, TestCaseType.MANUAL, scenario);
```

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java [hunk 14]
```diff
@@ -115,16 +141,34 @@ public void exportResults() throws IOException
...
+        parameters.setSteps(ManualStepConverter.convert(storyTitle, scenario.getTitle(), scenario.collectSteps()));
+        return parameters;
+    }
+
+    private CucumberTestCaseParameters createCucumberTestCaseParameters(Scenario scenario)
+    {
+        CucumberTestCaseParameters parameters = createTestCaseParameters(TestCaseType.CUCUMBER, scenario,
+                CucumberTestCaseParameters::new);
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/exporter/XrayExporter.java:156]:
```java
CucumberTestCaseParameters parameters = new CucumberTestCaseParameters();
fillTestCaseParameters(parameters, TestCaseType.CUCUMBER, scenario);
```

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

**Pull request merged**